README.md
# Virtual IP2SL (IP to Serial)

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=WREP29UDAMB6G)

Provides bidirectional TCP-to-serial access to physical serial ports connected to the
host running this microservice by emulating an iTach IP to Serial (IP2SL). Each instance
of the Virtual IP2SL microservice can expose up to eight physical RS232 serial ports.

I decided to build this after having physical USB to serial adapters hooked up to a
Raspberry Pi (and the native Raspberry Pi GPIO pin outs), but had several iOS and other client applications 
which supported RS232 over IP by implementing a subset of the published iTach TCP protocol.
While Open Source projects existed to emulate iTach IR devices, none implemented raw access
to serial ports via TCP. For my home, I paired this with [Flex IP](https://amazon.com/Global-Cache-iTach-Flex-IP/dp/B00C6FRPIC/?tag=carreramfi-20) hardware devices for cases where the serial port connection wasn't in the rack with my Raspberry Pi.

## Configuration

By default, the Virtual IP2SL is configured to open a single port attached 
at 9600 baud to one USB serial port adapter on /dev/ttyUSB0. However, a
wide variety of serial port configurations are possible, up to eight
serial ports per Virutal IP2SL instance. The environment variable IP2SL_CONFIG
can be used to point to YAML config files other than [config/default.yaml](config/default.yaml).

Below is an example of four USB serial ports connected.
See the "[iTach TCP API Specification](https://www.globalcache.com/files/releases/flex-16/API-Flex_TCP_1.6.pdf)"
PDF manual for the available configuration values for each serial port.

```yaml
serial:
  1: # port 1
    path: /dev/ttyUSB0
    baud: 9600
    flow: FLOW_NONE
    timeout: 4 # optional, default = 5 seconds

  2: 
    path: /dev/ttyUSB1
    baud: 9600
    flow: FLOW_HARDWARE

  3: 
    path: /dev/ttyUSB2
    baud: 14400

  4: 
    path: /dev/ttyUSB3
    baud: 115200
```

## Running

#### Standalone

```bash
python3 ip2sl
```

#### As Docker Container

Under Docker (may require modifying Dockerfile to specify the BUILD_FROM architecture base you need):

```bash
docker build -t virtual-ip2sl .
docker run virtual-ip2sl
```

NOTE: Since multicast is not supported from within a Docker container, the AMX discovery beacon will
not be published so you must manually configure clients to communicate with your Virtual IP2SL
instance.

#### As Home Assistant Hass.io Add-On

To run as a Home Assistant Hass.io add-on, install the repository:

 https://github.com/rsnodgrass/hassio-addons

## Network Ports

This microservice implements the open AMX Discovery Beacon protocol, raw TCP sockets to 
RS232 serial ports, and a TCP Port exposing the iTach command protocol.

The Virtual IP2SL listens on a variety of TCP ports, both for controlling the service
as well as the configuration for each serial port interface. Data sent to any of these
ports is relayed directly out the RS232 serial port associated with that TCP port in
configuration. Similarly, any data received from the RS232 will be written to the
TCP port.

| TCP Port | Description                              |
| -------- | ---------------------------------------- |
| 4998     | iTach TCP API command/control port       |
| 4999     | TCP port to the first serial port        |
| 5000     | ... second serial port *(optional)*      |
| 5001     | ... third serial port *(optional)*       |
| 5002     | ... fourth serial port *(optional)*      |
| 5003     | ... fifth serial port *(optional)*       |
| 5004     | ... sixth serial port *(optional)*       |
| 5005     | ... seventh serial port *(optional)*     |
| 5006     | ... eighth serial port *(optional)*      |

* For security, it is recommended disabling any ports that are not in use.
If no configuration exists for a given serial port (1-8), the associated TCP port
will not be opened.

## Example TTY Paths

The following are a variety of example TTY paths for different serial port interfaces:

| Serial Path                 | Description                                         |
| --------------------------- | --------------------------------------------------- |
| /dev/ttyS0                  | Raspberry Pi mini UART GPIO                         |
| /dev/ttyAMA0                | Raspberry Pi GPIO pins 14/15 (pre-Bluetooth RPi 3)  |
| /dev/serial0                | RPi 3/4 serial port alias 1                         |
| /dev/serial1                | RPi 3/4 serial port alias 2                         |
| /dev/tty.usbserial          | MacOS USB serial adapter                            |
| /dev/ttyUSB0                | Raspberry Pi USB serial adapter 1                   |
| /dev/ttyUSB1                | Raspberry Pi USB serial adapter 2                   |
| /dev/tty.usbserial-A501SGSU | StarTach ICUSB232I (8-port) serial port 1 (MacOS)   |
| /dev/tty.usbserial-A501SGSV | StarTach ICUSB232I (8-port) serial port 2 (MacOS)   |

## See Also

* [Virtual IP2SL for Home Assistant](https://github.com/rsnodgrass/hassio-addons/tree/master/virtual-ip2sl-addon)

#### Clients

* [iTest for Windows](https://www.globalcache.com/downloads/) and [iTest for MacOS (by Martijn Rijnbeek)](http://www.rmartijnr.eu/itest.html) - tools for connecting and sending test queries
* [Home Assistant IP2SL client by tinglis1](https://github.com/tinglis1/home-assistant-custom/tree/master/custom_components/notify) (last updated 2016)
* [Home Assistant gc100 sensor/switch (no serial)](https://www.home-assistant.io/components/gc100)

#### API

Special thanks to [Global Caché](https://www.globalcache.com/products/) for opening and publishing TCP control APIs:

* [iTach Flex TCP API Specification v1.6](https://www.globalcache.com/files/releases/flex-16/API-Flex_TCP_1.6.pdf)
  (earlier [v1.5 specificaiton](https://www.globalcache.com/files/docs/API-iTach.pdf))
* [iTach TCP/IP to Serial (RS232) specs](https://www.globalcache.com/products/itach/ip2slspecs/) and [Flex specs](https://www.globalcache.com/products/flex/flc-slspec/)

#### Related Emulators

* [GlobalCovfefe](https://platformio.org/lib/show/5679/GlobalCovfefe): Global Cache device emulator with one IR sender and one optional IR learner
* [ESP8266iTachEmulator](https://github.com/probonopd/ESP8266iTachEmulator): iTach/LIRC IP2IR emulator for IR signals
* [Global Cache Wireless Emulator (ESP8266)](https://hackaday.io/project/8233-global-cache-wireless-emulator-esp8266)

## Community Support

Links to several forums for community engagement around iTach Flex:

* https://community.home-assistant.io/t/itach-ip2sl/28805
* https://community.smartthings.com/t/home-theater-macro-global-cache-itach/126450

## TODO

NOTE: While this works for my use cases (and most common ones users will encounter),
it would be great to have other contributors help take this to the next level and
implement features, stability improvements, etc.

Planned:

* improve error handling (e.g. invalid port modules specified in get_SERIAL)

Ideas for eventual implementation (**feel free to contribute; no plans by me to add**):

* add support for RS485 connections
* persist set_SERIAL changes across restarts (optional)
* enhance security of the IP2SL interface (e.g. authentication token headers)
* web UI console showing details about the config and each port (including metrics)
* emulation compatibility for [GC-100-xx](https://www.globalcache.com/files/docs/API-GC-100.pdf)
* add an optional MQTT interface
* when Docker supports UDP multicast publishing from within a container, enable beacon for Docker image



CHANGELOG.md
# Virtual IP2SL (IP to Serial)

## 0.0.1 (2019-08-07)

Initial release


ip2sl/proxy.py
import logging

import os
import time
import serial
import select
import threading
import socketserver

import util
import ip2serial

log = logging.getLogger(__name__)

BUFFER_SIZE=4096

# initialize the map of serial port number to TCP port; we only limit to 8 ports
# since existing hardware which implements the Flex command protocol isn't found
# in sizes larger than 8 ports, so unclear what client behavior would be.
SERIAL_PORT_TO_TCP_PORT = { 8: 5007 }
def initialize_serial_port_to_tcp_port():
    num_ports = 8
    for i in range(1, num_ports):
        SERIAL_PORT_TO_TCP_PORT[i] = 4998 + i
initialize_serial_port_to_tcp_port()

Serial_Proxies = {}
def get_serial_proxies():
    return Serial_Proxies

""" 
Handler that proxies data to/from a specific serial port to the connected TCP
connection. This is instantiated once per client connection.  Since the serial
port communication is not multiplexed, only allow a single instance of this
instantiated at a time (this is the default behavior given the current threading model).
"""
class TCPToSerialProxy(socketserver.StreamRequestHandler):

    def __init__(self, request, client_address, server):
        log.debug(f"New serial connection from %s: %s", client_address[0], request)
        self._server = server
        self._client_id = client_address[0]
        self._running = True

        # call the baseclass initializer as the last thing; note __init__ waits on bytes to call handle()
        socketserver.StreamRequestHandler.__init__(self, request, client_address, server)

    def handle(self):
        tcp_client = self.request 

        ip2s = self._server._serial
        raw_serial = ip2s.get_raw_serial()
        serial_fd = raw_serial.fileno()
        tty_path = ip2s._tty_path

        check_if_running_delay = 2.0
        while self._running:
            read_ready, write_ready, exception = select.select([tcp_client, serial_fd],
                                                               [], [], check_if_running_delay)

            if tcp_client in read_ready:
                data = tcp_client.recv(BUFFER_SIZE)
                log.debug("Proxy %s --> %s: %s", self._client_id, tty_path, data)
                print(f"Proxy {self._client_id} --> {tty_path}: {data}")
                if raw_serial.write(data) <= 0:
                        break

            # if data available from the serial port, read it and forward to TCP socket
            if serial_fd in read_ready:
                time.sleep(0.05) # wait 50 ms for serial buffer to queue up
                data = raw_serial.read(raw_serial.in_waiting)
                log.debug("Proxy %s <-- %s: %s", self._client_id, tty_path, data)
                print(f"Proxy {self._client_id} <-- {tty_path}: {data}")
                if tcp_client.send(data) <= 0:
                        break

class IP2SLServer(socketserver.TCPServer):
    def __init__(self, server_address, RequestHandlerClass, serial_connection):
        socketserver.TCPServer.__init__(self, server_address, RequestHandlerClass)
        self._serial = serial_connection

"""Ensure all listeners are cleanly shutdown"""
def stop_all_listeners():
    for port_number, server in Serial_Proxies.items():
        stop_proxy(port_number)

def stop_proxy(port_number):
    log.debug(f"Stopping proxy for serial {port_number}")
    server = Serial_Proxies[port_number]
    if server != None:
       server.shutdown()
       server.server_close()

def start_proxy(port_number, serial_config, config):
    host = util.get_host(config)
    tcp_port = SERIAL_PORT_TO_TCP_PORT[port_number]

    log.info(f"Serial {port_number} (TCP port {host}:{tcp_port}) config: {serial_config}")
    serial_connection = ip2serial.IP2SLSerialInterface(serial_config)

    # FIXME: if serial port /dev/tty does not exist, should port be opened? or pending connection of device?
    # ...or a queue that is checked periodically and reopened?

    # each listener has a dedicated thread (one thread per port, as serial port communication isn't multiplexed)        
    log.info(f"Starting thread for TCP proxy to serial {port_number} at {host}:{tcp_port}")
    proxy = IP2SLServer((host, tcp_port), TCPToSerialProxy, serial_connection)
    
    proxy_thread = threading.Thread(target=proxy.serve_forever)
    proxy_thread.daemon = True # exit the server thread when the main thread terminates
    proxy_thread.start()

    # retain references to the thread and proxy
    Serial_Proxies[port_number] = ( proxy )
    return proxy

def start_serial_proxies(config):
    # start the individual TCP listeners for each serial port proxy
    for port_number, serial_config in config['serial'].items():
        port_number = int(port_number)
        start_proxy(port_number, serial_config, config)


ip2sl/beacon.py
import logging

import os
import re
import uuid
import time
import threading
import socket

import util

log = logging.getLogger(__name__)

MULTICAST_PORT = 9131
MULTICAST_IP   = '239.255.250.250'
MULTICAST_TTL  = int(os.getenv('IP2SL_BEACON_HOPS', 2)) # after TWO network hops the beacon packet should be discarded

# Implements a version of the AMX Beacon device discovery protocol with periodic heartbeats
class AMXDiscoveryBeacon():
    def __init__(self, config):
        self._config = config

        # heartbeat interval in seconds (default is every 10 seconds); ENV override for testing
        self._beacon_interval = max(1, int(os.getenv('IP2SL_BEACON_INTERVAL', '10')))

        self._console_host = util.get_host(config)
        self._console_port = int(os.getenv('IP2SL_CONSOLE_PORT', 4444))

        self._thread = threading.Thread(target=self.heartbeat, args=())
        self._thread.daemon = True
        self._thread.start()

    def get_mac(self):
        return ''.join(re.findall('..', '%012x' % uuid.getnode())).upper()

    def heartbeat(self):
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
        sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, MULTICAST_TTL)

        # iTach Flex discovery beacon is a AMX-styles multicast UDP packet sent to IP 239.255.250.250, port 9131.
        data = {
            'UUID'       : f"VirtualIP2SL_{self.get_mac()}", # required; unique identifer for this instance
            'SDKClass'   : 'Utility',            # required
            'Make'       : 'GlobalCache',        # required
            'Model'      : 'iTachFlexEthernet',  # required; note GC-100-12 for legacy model, or iTachIP2SL for v1.5 API
#            'Config-URL' : f"http://{self._console_host}:{self._console_port}",
            'Revision'   : '710-3000-18',
            'Pkg_Level'  : '', # "GCPK001",
            'PCB_PN'     : '025-0034-12',
            'Status'     : 'Ready'
            }
        heartbeat_packet = "AMXB" + ''.join(F"<-{k}={v}>" for (k,v) in data.items())

        while True:
            log.debug(f"Broadcasting heartbeat beacon to {MULTICAST_IP}:{MULTICAST_PORT}: {heartbeat_packet}")
            heartbeat_packet += "\r"
            sock.sendto(heartbeat_packet.encode(), (MULTICAST_IP, MULTICAST_PORT))
            time.sleep(self._beacon_interval)


ip2sl/ip2serial.py
import sys
import time
import serial
import logging
import threading

log = logging.getLogger(__name__)

DEFAULT_CONFIG = {
    'baud':      9600,
    'flow':      'FLOW_NONE',
    'parity':    'PARITY_NO',
    'stop_bits': 'STOPBITS_1',
    'timeout':   10
}

PARITY = {
    'PARITY_NO':   'N', # serial.PARITY_NONE,
    'PARITY_ODD':  'O', # serial.PARITY_ODD,
    'PARITY_EVEN': 'E'  # serial.PARITY_EVEN
}

STOP_BITS = {
    'STOPBITS_1':  1, # serial.STOPBITS_ONE
    'STOPBITS_2':  2  # serial.STOPBITS_TWO
}

FLOW_OR_DUPLEX = {
    'FLOW_HARDWARE': 'rs232',
    'FLOW_NONE':     'rs232',
    'DUPLEX_HALF':   'rs485',
    'DUPLEX_FULL':   'rs485'
}

def get_with_default(config, key):
    if key not in config:
        config[key] = DEFAULT_CONFIG[key] # update config with default if missing
    return config[key]

class IP2SLSerialInterface:
    def __init__(self, config):
        self._lock = threading.Lock()
        self._tty_path = config['path']
        self._config = config

        try:
            self._raw_serial = None
            self.get_raw_serial()
        except:
            log.error("Unexpected error: %s", sys.exc_info()[0])
            raise RuntimeError("Connect failure to {}".format(self._tty_path))

    def get_raw_serial(self):
        with self._lock:
            if self._raw_serial:
                return self._raw_serial
            
            config = self._config

            baud = int(get_with_default(config, 'baud'))
            self._baud = max(min(baud, 115200), 300) # ensure baud is ranged between 300-115200
            config['baud'] = self._baud # rewrite config to ensure it is within range

            self._parity    = PARITY[get_with_default(config, 'parity')]
            self._stop_bits = STOP_BITS[get_with_default(config, 'stop_bits')]
            self._timeout   = int( get_with_default(config, 'timeout') )
            
            # default to hardware flow control on (FLOW_HARDWARE)
            self._flow  = get_with_default(config, 'flow')
            flow_rtscts = (self._flow == 'FLOW_HARDWARE')
            flow_dsrdtr = flow_rtscts

            raw_serial = serial.Serial(self._tty_path,
                                         timeout=self._timeout,
                                         baudrate=self._baud,
                                         parity=self._parity,
                                         stopbits=self._stop_bits,
                                         bytesize=serial.EIGHTBITS,
                                         dsrdtr=flow_dsrdtr,
                                         rtscts=flow_rtscts)
            log.info(f"Connected to {self._tty_path} (config={self._config})")

            self._rs485 = self._flow in [ 'DUPLEX_FULL', 'DUPLEX_HALF' ]
            if self._rs485:
                message = f"RS485 not yet supported! (detected RS485 flow/duplex '{self._flow}' configuration)"
                log.error(message)
                raise RuntimeError(message)
                # FIXME: support RS485 --> ser.rs485_mode = serial.rs485.RS485Settings()

            # configure serial port for non-blocking mode (POSTIX only), requires timeout=0
            raw_serial.nonblocking()

            self._raw_serial = raw_serial

        return self._raw_serial


    def config(self):
        with self._lock:
            return self._config

    def close(self):
        with self._lock:
            raw_serial = self.get_raw_serial()
            if raw_serial:
                raw_serial.flush()
                raw_serial.close()

    # FIXME: this should really just change the existing serial connection, but for
    # now we will just swap and replace the serial object.
    def reset_serial_parameters(self, config):
        with self._lock:
            try:
                if self._raw_serial:
                    self._raw_serial.close()
                    self._raw_serial = None

                # update config and reopen raw serial
                self._config = config
                raw_serial = self.get_raw_serial()

            except:
                log.error("Unexpected error: %s", sys.exc_info()[0])
                raise RuntimeError("Connect failure to {}".format(self._tty_path))